src/content/topics/sdk/features/user-config.mdx

```diff
@@ -1764,7 +1764,7 @@ const context = {
 <CSTab label="Node.js SDK v7.x and later (TypeScript)">
 
 ```ts
-const context: ld.LDContext = {
+const context: ld.LDMultiKindContext = {
```

Comment:
I have concerns. What version of typescript did you use, and did you use exactly the sample, or slightly different?

Reply:
Because an LDContext is 

```  export type LDContext = LDUser | LDSingleKindContext | LDMultiKindContext;```

Reply:
I double checked with typescript 4.4.4 and it is working as expected.

Reply:
I checked with common, let me double check is js-sdk-client.

Reply:
Yeah, js-client-sdk has a test type for this, and also looks to be working fine.

Reply:
Demonstration. 

https://github.com/launchdarkly/LaunchDarkly-Docs/assets/4955475/2d7bbf8e-074e-413d-a543-0f4c93d4cb82

Type is happy, and IDE is happy. I compile it and it is fine.

Then I change it to something that is wrong, IDE shows an issue, compiler complains.

Reply:
I just tried typescript 5.1.6 and it also seems to be fine.

Reply:
The exact error message as well as typescript version would be most helpful.

Reply:
thanks for all the testing @kinyoklion

Reply:
Thanks Ryan, this came from troubleshooting the following ticket: https://launchdarklysupport.zendesk.com/agent/tickets/46240

I tested using the hello-node-typescript and essentially copy and pasted the customers multi context setup from the ticket and tried using that and got the following error:

```
TSError: ⨯ Unable to compile TypeScript
index.ts (39,7): Type '{ kind: string; user: LDSingleKindContext; account: LDSingleKindContext; }' is not assignable to type 'LDContext'.
 Type '{ kind: string; user: LDSingleKindContext; account: LDSingleKindContext; }' is not assignable to type 'LDMultiKindContext'.
  Types of property 'kind' are incompatible.
   Type 'string' is not assignable to type '"multi"'. (2322)
```

Changing it to LDMultiKindContext seemed to make the error go away in my own testing, but I realize this PR might've been over-eager/premature, since the customer tested this out in the JS client-side SDK and it didn't work for them

Reply:
I see problems here. A multi-kind context is NOT made of single kind contexts. So that will not work.

Reply:
If they want to type the interior parts of a multi-kind context, then they would be LDContextCommon.

Reply:
I would still ask them what typescript version.

Reply:
Well, I tried 3, and it is so broken it doesn't get that far.

Reply:
I see. AFAIK, I don't think we have any notice of minimum required typescript version in our docs. Perhaps a callout would be good?

Reply:
Sadly it isn't so straightforward. I would guess there are many people who are using typescript 3 and it is just fine. But it doesn't work with out test setup for sure. But it fails before it gets to any of our code because of dependencies.

Reply:
I would first address their problem with trying to make a multi-kind context with singles. The main difference is the components of a multi-kind context do not themselves have kinds (because the kind is in the map they are contained in).

Reply:
Yep, because that is even an older typescript. I did try that project with 3.x, and 4.x and it works.

Reply:
![image](https://github.com/launchdarkly/LaunchDarkly-Docs/assets/4955475/97452f24-2e74-4b67-8106-a5cf7bb35fa1)

Reply:
That project should be updated for sure, but it happens to work for a single kind, with a 6 year old version of typescript.

Reply:
https://github.com/launchdarkly/hello-node-typescript/pull/6

Reply:
@kinyoklion is there a way to do the multi context without defining the contexts within the definition of the multi context itself?

This worked for me on the Node.JS server-side SDK, but it didn't work for the customer using v3 of the JS client-side SDK:

```
const AccountContext: LaunchDarkly.LDContextCommon = { 
  key: '123' 
} 
const UserContext: LaunchDarkly.LDContextCommon = { 
  key: '456', 
} 
const multiContext: LaunchDarkly.LDContext = { 
  kind: "multi", 
  user: UserContext, 
  account: AccountContext 
}
```
I would expect it to work in both SDKs, so not sure where we're going wrong

Reply:
That should work in both SDKs. The types for contexts are basically the same between the two.

Reply:
But they are using TS 3, so let me try specifically that, so we may need to try specifically that with specifically 3.

Reply:
I tried it out with the client SDK, with typescript 3.9, and it worked. What is the error they are getting this time?

Reply:
Still the same error: ldclient.es.js:1 [LaunchDarkly] Invalid context specified. Please see https://docs.launchdarkly.com/sdk/client-side/javascript#initializing-the-client for instructions on SDK initialization.

Reply:
ack, my bad for the confusion... I was getting the type error, and they're getting that error.

The keys they're using in their latest example are just strings '123' and '345'... I would expect that to work just fine?

Reply:
Yeah, should. Let me test some combinations. The kinds are the kinds they are using in the ticket as well, correct?

Reply:
yes, `user` and `account` are the kinds the customer is using

Reply:
So, in their original code they had stuff like
```
              key: userData.Data.UserName,
```
Which, if that was an empty string, or undefined, then that context would not be valid.

This is the code that does the validation:
```
function checkContext(context, allowLegacyKey) {
  if (context) {
    if (allowLegacyKey && (context.kind === undefined || context.kind === null)) {
      return context.key !== undefined && context.key !== null;
    }
    const key = context.key;
    const kind = context.kind === undefined ? 'user' : context.kind;
    const kindValid = validKind(kind);
    const keyValid = kind === 'multi' || (key !== undefined && key !== null && key !== '');
    if (kind === 'multi') {
      const kinds = Object.keys(context).filter(key => key !== 'kind');
      return (
        keyValid &&
        kinds.every(key => validKind(key)) &&
        kinds.every(key => {
          const contextKey = context[key].key;
          return contextKey !== undefined && contextKey !== null && contextKey !== '';
        })
      );
    }
    return keyValid && kindValid;
  }
  return false;
}
```

Basically

1.) A kind must exist at the top level.
2.) A kind must be valid. ASCII letters numbers, dot, dash, and underscore. It also cannot be "kind".
3.) If the kind is "multi" then each top level field must be an object containing a key, and that key must be a non-empty string.
4.) If it isn't multi, then the kind must be valid, and must not be "kind".

Reply:
Maybe you can ask them to JSON.stringify the context they are setting and log it. Then verify that each context has a key and so forth. Or use data that they could send us.

Reply:
Yes thanks for all the guidance!